cip/CIP2016-06-14-Harmonize-comparability-orderability-equality-and-equivalence.adoc

```diff
@@ -0,0 +1,233 @@
+<=>
+a1 = b1 && a2 = b2 && ...
+----
+
+For clarity, we also repeat the current equality semantics of maps here (at the time of writing this CIP). Two maps `m1` and `m2` are considered equal if
+
+* `m1` and `m2` must have the same keys.
+* For each key `k`,
+** either `m1.k = m2.k` is `true`,
+** or both `m1.k IS NULL` and `m2.k IS NULL`
+
+This is at odds with the common interpretation of `null` as standing for any possible value.
+
+It is aligned though with the most common use case for maps with `null` entries which is to update multiple properties at once, e.g. `SET n += { n: 12, remove_this_key: null }`. In this case, there is no need to differentiate between different `null` values as `null` merely serves as a marker for keys to be removed. Current equality semantics makes it easy to check if two maps correspond to the same property update in this scenario.
+
+However this type of update map comparison is rare and could be emulated using a more complex predicate. The current rules do however break symmetry with how equality handles `null` in all other cases. This becomes more apparent by considering these two examples
+
+* `expr1 = expr2` => `null` if `expr1 IS null && expr2 IS NULL`
+* `{a: expr1} = {a: expr2}` => `true` (!!) if `expr1 IS null && expr2 IS NULL`
+
+To fix this, we propose that two maps `m1` and `m2` instead should be equal if
+
+* `m1` and `m2` have the same keys (the order of keys as returned by `keys` does not matter)
+* For each such key `k`, `m1.k = m2.k`
+
+As a consequence of these changes, plain equality is not reflexive in a classic sense (Consider: `{a: null} = {a: null}`, `[null] = [null]`). However this was already the case (Consider: `null = null` => `null`).
+
+However, Equality is reflexive for values that do not involve `null`.
+
+=== Equivalence ===
+
+Equivalence remains unchanged but now can be defined succinctly as being identical to equality except that:
+* Any two `null` values are treated as equivalent (both directly or inside nested structures).
+* Any two `NaN` values are treated as equivalent (both directly or inside nested structures).
+
+Note that orderability is aligned with equivalence, i.e. equivalent values have the same position under orderability.
+
+Equivalence is reflexive for all values.
+
+=== Summary of Conceptual Model
+
+This proposal aims to simplify the conceptual model around equality, comparison, sorting, and grouping:
+
+* Equality: Equality follows natural, literal equality. However, values involving `null` are never equal. Nested structures are first tested for equality by shape (keys, size) and then componentwise. This ensures that equality is compatible with interpreting `null` as "unknown" or "could be any value".
+* Comparability: Comparability is aligned with equality. Two values of the same type in the global sort order are always comparable. Two values of different types are always incomparable and values involving `null` are always incomparable.
```

Comment:
It would be nice if "global sort order" could link back to that section in the document in which this was defined